File: chapter02/main/src/main/java/game/Player.java
package game;

import java.util.ArrayList;
import java.util.List;

public class Player {

  private static int _ID = 1;
  private int userId;
  private List<Card> cards;
  private boolean isDrawing;
  private int money;
  private int moneyOnDesk;

  public Player() {
    this.userId = _ID++;
    cards = new ArrayList<>();
    isDrawing = true;
    money = 10; // 初始为石原里美
  }

  public void addMoney(int money) {
    this.money += money;
  }

  public int getMoney() {
    return money;
  }

  public int getUserId() {
    return userId;
  }

  public boolean isDrawing() {
    return isDrawing;
  }


  public List<Card> getCards() {
    return cards;
  }

  public int getCardSum() {
    int result = 0;
    int num_A = 0;
    for (Card temp : cards) {
      result += temp.getValue();
      if (temp.getValue() == 1) {
        num_A++;
      }
    }
    if (num_A >= 1) {
      if (result + 10 > 21) {
        return result;
      } else {
        return result + 10;
      }
    } else {
      return result;
    }
  }

  void addCard(Card card) {
    this.cards.add(card);
  }

  public void setStopDrawing() {
    isDrawing = false;
  }

  public void reset() {
    this.cards.clear();
    this.isDrawing = true;
  }

  public int getMoneyOnDesk() {
    return moneyOnDesk;
  }

  public void setMoneyOnDesk(int moneyOnDesk) {
    this.moneyOnDesk = moneyOnDesk;
  }

  public String showCards() {
    StringBuilder sb = new StringBuilder();
    for (Card card : getCards()) {
      sb.append(card);
      sb.append(" ");
    }
    if (!isDrawing) {
      sb.append("(").append(getCardSum()).append(")");
    }
    return sb.toString();
  }
}

File: chapter02/main/src/main/java/game/Player.kt
package game

import java.util.*

open class Player {
    val userId: Int = _ID++
    open val cards: MutableList<Card> = ArrayList()
    var isDrawing: Boolean = true
        private set
    var money: Int = 10
        private set
    var moneyOnDesk: Int = 0

    val cardSum: Int
        get() {
            return if (cards.count { it.value == 1 } == 0) {
                cards.sumBy { it.value }
            } else {
                val sum = cards.sumBy { it.value }
                when (sum) {
                    !in 0..21 -> sum
                    else -> sum + 10
                }
            }
        }

    fun addMoney(money: Int) {
        this.money += money
    }

    fun addCard(card: Card) {
        this.cards.add(card)
    }

    fun setStopDrawing() {
        isDrawing = false
    }

    fun reset() {
        this.cards.clear()
        this.isDrawing = true
    }

    open fun showCards(): String {
        val cardsInfo = cards.joinToString(separator = " ")
        return if (isDrawing) {
            cardsInfo
        } else {
            "$cardsInfo($cardSum)"
        }
    }

    companion object {
        private var _ID = 1
    }
}
